src/mlpack/core/tree/spill_tree/single_tree_traverser_impl.hpp

```diff
@@ -0,0 +1,138 @@
+/**
+ * @file single_tree_traverser_impl.hpp
+ * @author Ryan Curtin
+ * @author Marcos Pividori
+ *
+ * A nested class of SpillTree which traverses the entire tree with a
+ * given set of rules which indicate the branches which can be pruned and the
+ * order in which to recurse.  This traverser is a depth-first traverser.
+ */
+#ifndef MLPACK_CORE_TREE_SPILL_TREE_SINGLE_TREE_TRAVERSER_IMPL_HPP
+#define MLPACK_CORE_TREE_SPILL_TREE_SINGLE_TREE_TRAVERSER_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "single_tree_traverser.hpp"
+
+namespace mlpack {
+namespace tree {
+
+template<typename MetricType,
+         typename StatisticType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitMetricType, typename SplitMatType>
+             class SplitType>
+template<typename RuleType>
+SpillTree<MetricType, StatisticType, MatType, HyperplaneType, SplitType>::
+SingleTreeTraverser<RuleType>::SingleTreeTraverser(RuleType& rule) :
+    rule(rule),
+    numPrunes(0)
+{ /* Nothing to do. */ }
+
+template<typename MetricType,
+         typename StatisticType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitMetricType, typename SplitMatType>
+             class SplitType>
+template<typename RuleType>
+void SpillTree<MetricType, StatisticType, MatType, HyperplaneType, SplitType>::
+SingleTreeTraverser<RuleType>::Traverse(
+    const size_t queryIndex,
+    SpillTree<MetricType, StatisticType, MatType, HyperplaneType, SplitType>&
+        referenceNode)
+{
+  // If we are a leaf, run the base case as necessary.
+  if (referenceNode.IsLeaf())
+  {
+    for (size_t i = 0; i < referenceNode.NumPoints(); ++i)
+      rule.BaseCase(queryIndex, referenceNode.Point(i));
+  }
+  else
+  {
+    if(referenceNode.Overlap())
```

Comment:
Yes, you are right. I updated many files with the same situation in: https://github.com/mlpack/mlpack/pull/747/commits/e06cfc8f950430060506bc895a94f69b1f3fb29d